keps/sig-api-machinery/3352-aggregated-discovery/README.md

```diff
@@ -0,0 +1,788 @@
+How will we know that this has succeeded? -->
+
+### Non-Goals
+
+<!-- What is out of scope for this KEP? Listing non-goals helps to
+focus discussion and make progress. -->
+
+- Since the current discovery separated by group-version is already
+  GA, removal of the endpoint will be difficult. This KEP will solely
+  focus on introducing the new aggregated endpoint and will not cover
+  deprecation.
+
+## Proposal
+
+We are proposing a new endpoint `/discovery/v1` as an aggregated
+endpoint for all discovery documents. Discovery documents can
+currently be found under `apis/<group>/<version>` and `/api/v1` for
+the legacy group version. This discovery endpoint will support
+publishing an ETag so clients who already have the latest version of
+the aggregated discovery can avoid redownloading the document.
+
+We will add a new controller responsible for aggregating the discovery
+documents when a resource on the cluster changes. There will be no
+conflicts when aggregating since each discovery document is
+self-contained.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!-- What are the caveats to the proposal? What are some important
+details that didn't come across above? Go in to as much detail as
+necessary here. This might be a good place to talk about core concepts
+and how they relate. -->
+
+### Risks and Mitigations
+
+<!-- What are the risks of this proposal, and how do we mitigate?
+Think broadly. For example, consider both security and how this will
+impact the larger Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!-- This section should contain enough information that the specifics
+of your change are understandable. This may include API specs (though
+not always required) or even code snippets. If there's any ambiguity
+about HOW your proposal will be implemented, this is the place to
+discuss them. -->
+
+We will expose a endpoint `/discovery` that will support JSON, and
+protobuf and gzip compression. The endpoint will serve the aggregated
+discovery document for all types that a Kubernetes cluster supports.
+
+### API
+
+The contents of this endpoint will be an `APIGroupList`, which is the
+same type that is returned in the discovery document at `/apis`. The
+`APIGroupList` contains a list of `APIGroup` types which includes a
+list of
+[`GroupVersionForDiscovery`](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L1071)
+types. We will modify this `GroupVersionForDiscovery` type to include
+a list of
+[`APIResource`](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L1080).
+This list of `APIResource` is what is currently published at the
+`/apis/<group>/version` endpoint and is what will be aggregated into
+the new endpoint that we publish.
```

Comment:
I think we can omit storageVersionHash and just let the storage version migrator keep looking at the old place?

Reply:
What's the guideline that says whether we can omit something?  I'm a little surprised at the mindset of having a new API and saying "no no old client, you have to keep using it the old way".  Why is this new API not the best way for *any* client to get discovery vs the manual crawl?

Reply:
I don't think SVM wants to see cached data.

Reply:
Another reason the old version will never go away is that the aggregator will be reading the old version off of aggregated apiservers to produce the new cached version.

Would it be less confusing if we picked a new name rather than a new version?

Reply:
Is there no way to hit `/discovery/v1` and cache bust?  Or are we so committed to a controller design that that's the key outcome?

Now that I look, I don't see anything in the KEP that bounds the latency of the doc - could we specify that in the design (we accept X latency in the absence of errors)?

Your point about SVM not wanting to see cached results... means that someone who is *mostly* ok with current results but occasionally needs to force that has no tool for it?  If SVM has a previous version of the results, and the aggregated APIs are down, as long as the new entry point can bound the cached endpoint on the etag, or omit it and get "the most recently bounded copy" (if it were seconds), is that really a problem for SVM?  Could we simply let the request for etag wait for the next controller interval to run (if short) such that it was always 'most recent / bounded'?

There seems to be a lot of expectations in the design of this that are focused on kubectl, but that's just one of our clients, and we *did* say "all clients" in the KEP...

Reply:
After chatting - the analogy here is webhooks and how we fail closed if we can't refresh our config in a bounded interval.  Clients of this API can assume it is slightly delayed, but if something goes wrong in the controller they have no way of knowing how delayed, which (for the namespace example, is potentially catastrophic).  I don't think that's a requirement for alpha, but it's potentially a requirement for beta to ensure clients who may have ambiguous use cases don't a) accidentally use this api and regress, b) identify how we solve for it.